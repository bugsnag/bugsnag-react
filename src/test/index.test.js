/* global jest, expect, it, beforeEach, test */

import React from 'react'
import renderer from 'react-test-renderer'
import index from '../'

class BugsnagReport {
  updateMetaData () {
    return this
  }
}

const bugsnag = {
  BugsnagReport,
  notify: jest.fn()
}

bugsnag.BugsnagReport.getStacktrace = jest.fn()

const ErrorBoundary = index(React).init(bugsnag)

beforeEach(() => {
  bugsnag.notify.mockReset()
})

test('formatComponentStack(str)', () => {
  const str = `
  in BadButton
  in ErrorBoundary`
  expect(index.formatComponentStack(str))
    .toBe(`in BadButton\nin ErrorBoundary`)
})

const BadComponent = () => {
  throw Error('BadComponent')
}

const GoodComponent = () => 'test'

it('renders correctly', () => {
  const tree = renderer
    .create(<ErrorBoundary><GoodComponent /></ErrorBoundary>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly on error', () => {
  const tree = renderer
    .create(<ErrorBoundary><BadComponent /></ErrorBoundary>)
    .toJSON()
  expect(tree).toBe(null)
})

it('calls notify on error', () => {
  renderer
    .create(<ErrorBoundary><BadComponent /></ErrorBoundary>)
    .toJSON()
  expect(bugsnag.notify).toHaveBeenCalledTimes(1)
})

it('does not render FallbackComponent when no error', () => {
  const FallbackComponent = jest.fn(() => 'fallback')
  const tree = renderer
    .create(<ErrorBoundary FallbackComponent={FallbackComponent}><GoodComponent /></ErrorBoundary>)
    .toJSON()
  expect(tree).toMatchSnapshot()
  expect(FallbackComponent).toHaveBeenCalledTimes(0)
})

it('renders FallbackComponent on error', () => {
  const FallbackComponent = jest.fn(() => 'fallback')
  const tree = renderer
    .create(<ErrorBoundary FallbackComponent={FallbackComponent}><BadComponent /></ErrorBoundary>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('passes the props to the FallbackComponent', () => {
  const FallbackComponent = jest.fn(() => 'fallback')
  renderer
    .create(<ErrorBoundary FallbackComponent={FallbackComponent}><BadComponent /></ErrorBoundary>)
  expect(FallbackComponent).toBeCalledWith({
    error: expect.any(Error),
    info: { componentStack: expect.any(String) }
  }, {})
})

it('it passes the beforeSend function to the Bugsnag notify call', () => {
  const beforeSend = () => {}
  renderer
    .create(<ErrorBoundary beforeSend={beforeSend}><BadComponent /></ErrorBoundary>)
    .toJSON()
  expect(bugsnag.notify).toBeCalledWith(
    expect.any(BugsnagReport),
    expect.objectContaining({ beforeSend: beforeSend })
  )
})
