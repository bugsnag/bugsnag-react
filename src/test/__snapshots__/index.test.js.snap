// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render FallbackComponent when no error 1`] = `"test"`;

exports[`renders FallbackComponent on error 1`] = `"fallback"`;

exports[`renders correctly 1`] = `"test"`;
